### PR TITLE
Support package and module in config file

### LINKED ISSUE #9883: Add `packages` and `modules` keys to config file format
**Feature**

Offer configuration variables for `packages` and `modules` to match the command-line flags, much like `files` matches listing files on the command line.  Intended to work exactly like the command line flags, with the exception that they would not override a `files` key and having `files` and either of the new keys in the same config file would create a conflict.

Usage meant to be something like this (actual code this would work with linked below):

```ini
[mypy]
python_version = 3.7
packages=spack,builtin.packages,builtin_mock.packages
mypy_path=bin,lib/spack,lib/spack/external,var/spack/repos/builtin
```

To match a command line like `mypy -p spack -p builtin.packages -p builtin_mock.packages`

**Pitch**

The main use comes from, [spack](GitHub.com/llnl/spack), a project I'm adding type checking to.  It's a package manager where the libraries, tool and all package definition files are written in python.  Having mypy check the library without arguments is simple, I can just write a `files` key that covers all files and it will work fine, but the packages are in a structure like this:

```
./var/spack/repos/
├── builtin
│   ├── packages
│   │   ├── 3proxy
│   │   │   └── package.py
│   │   ├── abduco
│   │   │   └── package.py
│   │   ├── abi-compliance-checker
│   │   │   └── package.py
... etc.
```

If I use a `files` key for these, they are all treated as duplicate `package` modules, and mypy errors out (probably appropriately).  They can however be treated as a package, and mypy works quite well on the library and the whole package tree in that mode (especially after the recursive checking fixes).  The trick now is that I can't make running `mypy` check the packages tree without arguments or wrapping or similar.

I'd like to add support for specifying these in the configuration file to bring them up to the same level of support as listing files, as mentioned in PR #9834.

### LINKED ISSUE #10728: Support package and module in config files
**Feature**

Allow listing modules/packages to check, within the MyPy config files.

**Pitch**

Today, it is possible to use the `files` config item (in mypy.ini, pyproject.toml, etc.) to tell MyPy what files to check.  You can also do this on the command-line, by simply listing the files.

If you want MyPy to check an entire module or package, you can do so on the command line (with `-m` or `-p`), but there is no way to do this in a config file.  Since it is possible to set `files` in config files, I think it should also be possible to set `module` and `package`.  (I am suggesting not-plural names "module" and "package" to match the command-line arguments).

One question you might ask is "Why don't you just list the directory containing your modules/packages?"  In my case, doing so gives me an error like this:

```
src/namespace1/namespace2/package1/file.py: error: Source file found twice under different module names: `package2` and `namespace1.namespace2.package2`
```

… or an error like this:

```
src/namespace1/namespace2/package1/__init.py__:NN: error: Skipping analyzing "namespace1": found module but no type hints of library stubs
```

All of my code is in the `src` directory.  In my `pyproject.toml`, I have `namespace_packages = true` and `mypy_path = ["src"]`.  The above error only happens if I try to specify a file or directory to check.  If instead I use `-p namespace1` (or `-p namespace1.namespace2`), everything is located, loaded, and typechecked successfully.

### PR DESCRIPTION
**Pull request by sameer-here**
### Description

Closes #10728

Specifying module(s) to check was supported only via CLI, this PR adds
the ability to specify a module from a config file.

## Test Plan

This is a pretty small change, I tested it by creating a `mypy.ini` file and verifying that adding the new options leads to expected outcomes (for example, specifying a `package` and `files` together will error here https://github.com/python/mypy/blob/ea933268427737976ef819ea3fff0599264e3d5b/mypy/main.py#L957-L960).  If there's a unit test to which I should contribute, I'd be happy to.

Do I need to rebuild the documentation after changing `config_file.rst`? If so, how is that done?

Thanks!

### CODE REVIEW

**Review by hauntsaninja — COMMENTED** [review]
Thanks, this looks reasonable. Can we make `packages` and `modules` plural everywhere for consistency with `files`?

**Comment by sameer-here** [comment]
@hauntsaninja looks like all checks are passing, is anything else needed?

File: docs/source/config_file.rst [hunk 1]
```diff
@@ -191,6 +191,22 @@ section of the command line docs.
 
     This option may only be set in the global section (``[mypy]``).
 
+.. confval:: packages
+
+    :type: comma-separated list of strings
+
+    A comma-separated list of packages which should be checked by mypy if none are given on the command
+    line. This option may only be set in the global section (``[mypy]``).  This option may be used
```
Comment (hauntsaninja) [inline docs/source/config_file.rst:199]: package isn't interchangeable with modules, right? Only packages will recursively type check submodules. E.g., maybe use the language from the command line docs?

Comment (sameer-here) [reply docs/source/config_file.rst]: You're correct, I've copied the language from the command-line docs.

File: mypy/options.py [hunk 2]
```diff
@@ -215,6 +215,13 @@ def __init__(self) -> None:
         # supports globbing
         self.files: list[str] | None = None
 
+        # A comma-separated list packages for mypy to type check
+        self.packages: list[str] | None = None
+
+        # A comma-separated list modules for mypy to type check;
```
Comment (hauntsaninja) [inline mypy/options.py:221]: Almost there, you missed updating this comment :-)

Comment (sameer-here) [reply mypy/options.py]: Removed the misleading comment- you've been very patient!

File: mypy/options.py [hunk 3]
```diff
@@ -218,8 +218,7 @@ def __init__(self) -> None:
         # A comma-separated list packages for mypy to type check
         self.packages: list[str] | None = None
 
-        # A comma-separated list modules for mypy to type check;
-        # This option can be used interchangably with package
+        # A comma-separated list modules for mypy to type check
```
Comment (JelleZijlstra) [inline mypy/options.py:221]:
```suggestion
        # A comma-separated list of modules for mypy to type check
```

File: mypy/options.py [hunk 4]
```diff
@@ -215,6 +215,12 @@ def __init__(self) -> None:
         # supports globbing
         self.files: list[str] | None = None
 
+        # A comma-separated list of packages for mypy to type check
+        self.packages: list[str] | None = None
+
+        # A comma-separated list of modules for mypy to type check
+        self.modules: list[str] | None = None
```
Comment (hauntsaninja) [inline mypy/options.py:222]:
```suggestion
        # A list of packages for mypy to type check
        self.packages: list[str] | None = None

        # A list of modules for mypy to type check
        self.modules: list[str] | None = None
```

**Review by hauntsaninja — APPROVED** [review]
Thanks for working on this!

**Pull request merged**